### PR TITLE
Histogram calculation with mask

### PR DESCRIPTION
**Pull request by yannlabou**
Related to #306

### CODE REVIEW

File: src/image_function.cpp [hunk 1]
```diff
@@ -747,6 +747,30 @@ namespace Image_Function
         }
     }
 
+    std::vector < uint32_t > Histogram( const Image & image, const Image & mask )
+    {
+        const Image maskedImage = BitwiseAnd( image, mask );
```
Comment (ihhub) [inline src/image_function.cpp]: The idea behind mask is to do not calculate some areas of image, meaning to do not include results from these areas of image. If we use bitwise logical AND the number of results in histogram remains the same. What we need is to add pixel values from original image only on areas where mask has non 0 values. For example, we found a circle on an image and we want a histogram only this circle so we use mask for this.

Comment (ihhub) [reply src/image_function.cpp]: The code in new function should be something like this:
```cpp
        const uint32_t rowSize = image.rowSize();
        const uint32_t rowSizeMask = mask.rowSize();

        const uint8_t * imageY     = image.data() + y * rowSize + x;
        const uint8_t * imageYMask = mask.data() + yMask * rowSizeMask + xMask;
        const uint8_t * imageYEnd  = imageY + height * rowSize;

        for( ; imageY != imageYEnd; imageY += rowSize, imageYMask += rowSizeMask ) {
            const uint8_t * imageX     = imageY;
            const uint8_t * imageXMask = imageYMask;
            const uint8_t * imageXEnd  = imageX + width;

            for( ; imageX != imageXEnd; ++imageX, ++imageXMask ) {
                if( (*imageXMask ) > 0 )
                    ++histogram[*imageX];
            }
        }
```

**Comment by ihhub** [comment]
@yannlabou shall I change the code based on my suggestions or wait for you? Additionally, we need unit tests for this function.

**Comment by yannlabou** [comment]
Feel free if you need it. I'm unfortunately on business trip and short on time this week.

**Comment by ihhub** [comment]
> Feel free if you need it. I'm unfortunately on business trip and short on time this week.

Got it. Have a fruitful trip! ;)

**Comment by ihhub** [comment]
Base changes are done. We need to add unit tests for this. To simplify work I'm proposing to add unit tests in such way: generate uniform image. Generate mask image. In mask image make a half of the image white (255) and another half black (0). Then the total result would be the same as in original Histogram but the histogram would have in 2 times less points.

**Comment by yannlabou** [comment]
@ihhub Thank you for finishing the PR. I wish I had more time these days.

**Comment by ihhub** [comment]
That's fine. It's still better than nothing ;)

**Pull request merged**